Amp Img

sareh opened the pull request:
Resolves #929

Adding an AmpImg component that extends Img. 

Integrating it into FigureContainer.

**Testing notes**
Visiting a canonical page renders images with `<img>`, with attributes `alt`, `src`, `height` and `width`.
http://localhost:7080/news/articles/c9rpqy7pmypo
<img width="247" alt="screen shot 2018-12-18 at 09 44 31" src="https://user-images.githubusercontent.com/3028997/50145455-8c6a0e00-02a9-11e9-8d02-4c69d799d2db.png">


Visiting an amp page renders images with `<amp-img>`, with attributes `alt`, `src`, `height` and `width`.
http://localhost:7080/news/articles/c9rpqy7pmypo.amp

<img width="266" alt="screen shot 2018-12-18 at 09 42 45" src="https://user-images.githubusercontent.com/3028997/50145457-8c6a0e00-02a9-11e9-9a42-567110239318.png">

Note, to run the validator and check everything is correct, in one window run:
```
npm run build && npm run start
```
In the second window run: 

```
amphtml-validator http://localhost:7080/news/articles/c9rpqy7pmypo.amp || true
```

(Note: the `npm run test:ampValidate` seems to be failing on the wait command.  Separate issue to fix this: https://github.com/BBC-News/simorgh/issues/1109 )
- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

sareh committed:
Update Image component to include AmpImg, which extends the canonical Image component

sareh committed:
Use Img in Figure Container

ChrisBAshton commented on src/app/components/Figure/Image/README.md:
````diff
@@ -0,0 +1,23 @@
+const WrappingContainer = ({ alt, height, src, width }) => (
+  <AmpImg alt={alt} src={src} height={height} width={width} />
+);
+```
````
Needs the rest of the README structure (guess I could have raised an issue for that!), e.g. https://github.com/BBC-News/psammead/tree/latest/packages/components/psammead-paragraph

sareh replied:
The CoC and other links will need to change when we move it to Psammead. I've added a note to the description in this Issue: https://github.com/BBC-News/psammead/issues/203

ChrisBAshton commented on src/app/components/Figure/Image/README.md:
```diff
@@ -0,0 +1,23 @@
+import { AmpImg } from '../components/Image';
+
+const WrappingContainer = ({ alt, height, src, width }) => (
+  <AmpImg alt={alt} src={src} height={height} width={width} />
```
```suggestion
  <AmpImg alt={alt} src={src} height={height} width={width} layout="responsive" />
```

sareh committed:
Use Image in Figure Container

ChrisBAshton commented on src/app/components/Figure/Image/fixtureData.js:
```diff
@@ -0,0 +1,16 @@
+  'https://ichef.bbci.co.uk/news/660/cpsprodpb/114FE/production/_104801907_79010.jpg';
+
+export const imageAltCustom = 'By Elisa Decker, from her series "Sidewalk"';
+export const imageSrcCustom =
```
Could we restructure this more like:

```js
export default {
  landscape: {
    alt: "Birthday illustration with llama and dogs",
    src: "https://ichef.bbci.co.uk/news/1024/cpsprodpb/537D/production/_104837312_crazyrichbirthday_illustration_dogs-nc976.jpg",
  }
  ..etc
};
```

sareh replied:
Done.

ChrisBAshton commented on src/app/components/Figure/Image/index.amp.jsx:
```diff
@@ -0,0 +1,6 @@
+import React from 'react';
+import { Img } from '.';
+
+const AmpImg = () => <Img as="amp-img" layout="responsive" />;
```
Nice 👍 

EDIT: we should still make it possible to override `layout`, just default it to `responsive`

sareh replied:
Okay, I didn't realise that we'd need that for this initial version. If we're going to make it a prop, which other values do you think we'll need to use? `intrinsic` and `fixed`? https://www.ampproject.org/docs/design/responsive/control_layout#the-layout-attribute

I'll want to add a unit test to capture the variance in this.

ChrisBAshton replied:
Ha - fair point! After looking at all the possible attributes we _could_ add - which feels like a premature optimisation - I'm now thinking we should keep it as hard-coded `layout="responsive"`.

EDIT - we don't even need to explicitly set the `layout`, as [AMP infers the layout behaviour from the width/height attributes](https://www.ampproject.org/docs/design/responsive/control_layout#what-if-the-layout-attribute-isn%E2%80%99t-specified?). So I'm now thinking we leave it out entirely, until it's needed!

ChrisBAshton commented on src/app/components/Figure/Image/index.amp.stories.jsx:
```diff
@@ -0,0 +1,27 @@
+  .add('16:9 landscape image', () => (
+    <AmpImg alt={imageAltLandscape} src={imageSrcLandscape} />
+  ))
+  .add('16:9 portrait image', () => (
```
```suggestion
  .add('9:16 portrait image', () => (
```

sareh committed:
Add tests for default export/import

ChrisBAshton commented on src/app/components/Figure/Image/index.amp.test.jsx:
```diff
@@ -0,0 +1,52 @@
+    <AmpImg
+      alt={imageAltPortrait}
+      src={imageSrcPortrait}
+      height={723}
```
Is this not `1024`? (if not, it's not `9:16` so ignore my comment above)

ChrisBAshton commented on src/app/components/Figure/Image/index.amp.test.jsx:
```diff
@@ -0,0 +1,52 @@
+      alt={imageAltCustom}
+      src={imageSrcCustom}
+      height={660}
+      width={660}
```
If we're doing a custom dimensions test, can we make it different to the square one?

sareh committed:
Merge branch 'latest' into amp-img

ChrisBAshton reviewed (commented):
Added some comments

sareh committed:
Refactor tests to remove duplication

sareh committed:
Correct image dimensions and examples

sareh committed:
Correct image dimensions and examples & refactor fixtureData exports

sareh committed:
Use withComponent syntax for AmpImg extension

Add conditional logic for Img AmpImg rendering in Figure container

sareh committed:
Refactor tests & stories & update snapshots

sareh committed:
Update Figure Container to include height and width properties

sareh committed:
Use platform context in figure container

sareh committed:
Merge branch 'latest' into amp-img

sareh committed:
Use snapshotTests shared file for amp tests

sareh committed:
Moving helpers to testHelpers directory

sareh commented:
Visiting a canonical page renders images with `<img>`, with attributes alt, src, height and width.
http://localhost:7080/news/articles/c9rpqy7pmypo

Visiting an amp page renders images with `<amp-img>`, with attributes alt, src, height and width.
http://localhost:7080/news/articles/c9rpqy7pmypo.amp

sareh commented:
For readability of our test files, I have not refactored all of the tests.

sareh commented on src/app/components/Figure/Image/index.amp.jsx:
```diff
@@ -0,0 +1,6 @@
+import 'styled-components';
+import { StyledImg } from '.';
+
+const AmpImg = StyledImg.withComponent('amp-img');
```
V4 of Styled Components adds the `as` prop https://www.styled-components.com/docs/api#as-polymorphic-prop. We should use this in future. 

However, since Simorgh is still using V3 of Styled Components, we cannot yet use the `as` prop and have to use the `withComponent` syntax instead.

dr3 replied:
Good news! https://github.com/BBC-News/simorgh/blob/latest/package.json#L93

ChrisBAshton replied:
Fair point - should we move to Psammead first and make use of latest v4 functionality, rather than update in Simorgh and then upgrade in the transfer?

sareh replied:
The `withComponent` syntax still works in v4, so it'll work in Psammead.

dr3 commented on package.json:
```diff
@@ -132,7 +132,8 @@
       "!**/src/app/helpers/tests/**",
-      "!**/*.stories.jsx"
+      "!**/*.stories.jsx",
+      "!**/testHelpers/*.{js,jsx}"
```
```suggestion
      "!**/testHelpers/**"
```

dr3 commented on src/app/components/Figure/Image/index.jsx:
```diff
@@ -1,8 +1,23 @@
   width: 100%;
 `;
 
+export const Img = ({ alt, height, src, width }) => (
```
Should this file not also be exporting AmpImg?

ChrisBAshton replied:
Yes I think that's what we agreed 👍

sareh replied:
Yes, this is fixed now.

ChrisBAshton replied:
```suggestion
export const Img = ({ alt, src, height, width }) => (
```

(ordering consistency)

sareh committed:
Update readme to reflect implementation

sareh committed:
Update export for AmpImg

sareh commented on src/app/components/Figure/Image/index.jsx:
```diff
@@ -1,8 +1,26 @@
+};
+
+// eslint-disable-next-line global-require
+export const AmpImg = require('./index.amp');
```
This isn't working in conjunction with  `import { AmpImg } from '../../components/Figure/Image';`

sareh committed:
Moving StyledImg to another file to remove circular dependencies

Fixes Amp page rendering

sareh committed:
Refactor AmpImg to use styled-components 'as' polymorphic prop to use '<amp-img>'

Add srcset prop to Img and AmpImg, ensuring it is optional

ChrisBAshton commented on src/app/components/Figure/Image/README.md:
````diff
@@ -0,0 +1,23 @@
+```jsx
+import Image from '../components/Image';
+
+const WrappingContainer = ({ alt, height, src, width }) => (
````
```suggestion
const WrappingContainer = ({ alt, src, height, width }) => (
```

(makes it consistent with the other below)

ChrisBAshton commented on src/app/components/Figure/Image/index.jsx:
```diff
@@ -1,8 +1,20 @@
+Img.propTypes = {
+  alt: string.isRequired,
+  height: number.isRequired,
+  src: string.isRequired,
```
Let's reorder these as well ^

ChrisBAshton commented on src/app/components/Figure/Image/index.jsx:
```diff
@@ -1,8 +1,20 @@
+  width: number.isRequired,
+};
+
+const Image = Img;
```
We can just `export default Img`, instead of making an intermediary - right?

ChrisBAshton commented on src/app/components/Figure/Image/index.stories.jsx:
```diff
@@ -1,12 +1,4 @@
-import React from 'react';
-import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
-import Image from './index';
+import { Img } from './index';
```
```suggestion
import { Img } from '.';
```

ChrisBAshton commented on src/app/containers/Figure/index.jsx:
```diff
@@ -1,37 +1,59 @@
-  copyright: string,
   captionBlock: objectOf(any),
+  copyright: string,
+  height: number.isRequired,
```
let's move height down for ordering

sareh replied:
The aim is to alphabetise the props, for readability.

ChrisBAshton replied:
Hmm - I see why you did that. But I think it's like CSS properties, we ought to group by type rather than alphabetically:

![45% CSS devs write by type vs 14% alphabetical](https://css-tricks.com/wp-content/uploads/2012/08/orgchart.png)

ChrisBAshton reviewed (changes requested):
Some changes requested - good start!

sareh committed:
Add srcset prop to ImageContainer

Remove conditional logic for originCode - it is now always expected to be passed by the CMS

sareh committed:
Add srcset prop to Figure container, to pass to Image component

sareh committed:
Reorder props, ensure height is required for AmpImg, tidy up imports/exports

sareh committed:
Merge branch 'latest' into amp-img

sareh commented:
The `amp-img` does not scale correctly. I've tried several different cases:

**Option 1**
`layout: 'responsive'` `height` & `width` defined. The docs say the height and width are only used for ratios, however, when used in Simorgh they get converted into px, and there is an overflow: 
<img width="1139" alt="screen shot option 1" src="https://user-images.githubusercontent.com/3028997/50280074-71330600-0443-11e9-966f-8c2196cc1d65.png">

**Option 2**
https://www.ampproject.org/docs/reference/components/amp-img#the-difference-between-responsive-and-intrinsic-layout
Have used `layout: "intrinsic"` as well, just to see whether it'd make a difference. It does not.

It appears that regardless of the layout value passed, the `fixed` layout is assumed. 
<img width="1138" alt="screen shot of fixed" src="https://user-images.githubusercontent.com/3028997/50279313-3e880e00-0441-11e9-81c9-02a1f9b2410d.png">

**Option 3**
https://www.ampproject.org/docs/reference/components/amp-img#scaling-an-image-up-to-a-maximum-width
Scaling up to a maximum width with a wrapper around the `<amp-img>` - e.g. with a 
```
const StyledWrapper = styled.div`
    max-width: 48rem;
```
This does constrain the image correctly, however the image width won't smoothly transition to be 6 out of 10 columns of our grid. 

**Option 4**
Am now exploring the `sizes` option: https://www.ampproject.org/docs/design/responsive/art_direction#sizes

sareh committed:
Reorder props to group height width & ratio.

Don't use ImagePlaceholder and Copyright for AmpImg

Use React camelcased srcSet for attribute

sareh committed:
Merge branch 'latest' into amp-img

sareh committed:
Merge branch 'amp-img' of github.com:bbc/simorgh into amp-img

sareh committed:
Removing styles from AmpImg & refactoring

sareh committed:
Update snapshots & fix ordering

sareh committed:
Add attribution attribute to amp-img

sareh committed:
Add tests for layout props for AmpImg & update storybook stories to not pass in height

sareh committed:
Move global styles decorator into CanonicalDecorator component

sareh committed:
Use CanonicalDecorator for Img and AmpDecorator for AmpImg

sareh commented:
Due to https://github.com/BBC-News/simorgh/pull/1104 being merged in, the Image component is only in Psammead now. 
This PR is not relevant to the initial issue #929, but will be to this one #953, after this Psammead PR is merged in: https://github.com/BBC-News/psammead/pull/241
